core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,163 @@
+   *
+   * @param a value to be written to the `Ref`
+   * @return `IO[E, Unit]`
+   */
+  def setAsync(a: A): IO[EA, Unit]
+
+  /**
+   * Apply a partial function `pf` to the content of the Ref, if `pf` is not defined over the input it will fail on `E` with `None`
+   */
+  final def collect[C](pf: PartialFunction[B, C]): ZRef[EA, Option[EB], A, C] =
+    new ZRef[EA, Option[EB], A, C] {
+      def set(a: A): IO[EA, Unit] = self.set(a)
+      def get: IO[Option[EB], C] = self.get.foldM(
+        error => ZIO.fail(Some(error)),
+        value => pf.lift(value).fold[IO[Option[EB], C]](ZIO.fail(None))(ZIO.succeed(_))
+      )
+      def setAsync(a: A): IO[EA, Unit] = self.setAsync(a)
+    }
+
+  /**
+   * Replace the content of the Ref with `c`
+   */
+  final def as[C](c: C): ZRef[EA, EB, A, C] = map(_ => c)
+
+  /**
+   * Transforms elements that are set into this Ref with a function.
+   */
+  final def contramap[C](ca: C => A): ZRef[EA, EB, C, B] =
+    new ZRef[EA, EB, C, B] {
+      def set(c: C): IO[EA, Unit]          = self.set(ca(c))
+      def get: IO[EB, B]                   = self.get
+      def setAsync(c: C): zio.IO[EA, Unit] = self.setAsync(ca(c))
+    }
+
+  /**
+   * Applies a filter to the content of the Ref. If the content of the Ref does not
+   * pass the filter, it will fail in `EB` with `None`
+   */
+  final def filter(f: B => Boolean): ZRef[EA, Option[EB], A, B] =
+    collect { case b if f(b) => b }
+
+  /**
+   * Transforms elements that are retrieved from this Ref with a function.
+   */
+  final def map[C](bc: B => C): ZRef[EA, EB, A, C] =
+    new ZRef[EA, EB, A, C] {
+      def set(a: A): IO[EA, Unit]          = self.set(a)
+      def get: IO[EB, C]                   = self.get.map(bc)
+      def setAsync(a: A): zio.IO[EA, Unit] = self.setAsync(a)
+    }
+
+  /**
+   * Maps from `E` error type to another.
+   */
+  final def mapError[EC](f: EA => EC): ZRef[EC, EB, A, B] = new ZRef[EC, EB, A, B] {
+    def set(a: A): IO[EC, Unit]          = self.set(a).mapError(f)
+    def get: IO[EB, B]                   = self.get
+    def setAsync(a: A): zio.IO[EC, Unit] = self.setAsync(a).mapError(f)
+  }
+
+  /**
+   * Discard the content of the Ref and replace it with `Unit`
+   */
+  final def unit: ZRef[EA, EB, A, Unit] = as(())
+
+  private[zio] def unifyError[E](ea: EA => E, eb: EB => E): ZRef[E, E, A, B] =
+    new ZRef[E, E, A, B] {
+      def set(a: A): IO[E, Unit]          = self.set(a).mapError(ea)
+      def setAsync(a: A): zio.IO[E, Unit] = self.setAsync(a).mapError(ea)
+      def get: IO[E, B]                   = self.get.mapError(eb)
+    }
+
+  /**
+   * Atomically modifies the `Ref` with the specified function. This is not
+   * implemented in terms of `modify` purely for performance reasons.
+   *
+   * @param f function to atomically modify the `Ref`
+   * @return `IO[E, B]` modified value of the `Ref`
+   */
+  def update[E](f: B => A)(implicit ea: EA <:< E, eb: EB <:< E): IO[E, B] = {
```

Comment:
@jdegoes @regiskuckaertz , is it what you had in mind ? tests seem to pass, not sure about this encoding

Reply:
We have to override `update` and `modify` inside the `ZRef.make` constructor for the shortcut versions. Here, while we know the _types_ are equal, we don't know that they are "original" — they may have been mapped many times. So it's not safe to use the unsafe methods.